### PR TITLE
TypeError: parameter.value.getTime is not a function

### PR DESCRIPTION
**Pull request by atsage**
Fixed an error TypeError: parameter.value.getTime is not a function.  Parameter needs to be converted to a date to avoid errors.

### CODE REVIEW

**Comment by atsage** [comment]
There are a few issues fixed with this PR:

https://github.com/patriksimek/node-mssql/issues/377
https://github.com/patriksimek/node-mssql/issues/185
https://github.com/tediousjs/tedious/issues/608

I've created a node that depends on node-mssql and so my node also has these datetime errors.

I also notice that the node on npm is not the same code that's in this repository.  How often do you publish to npm?

File: src/data-type.js [hunk 1]
```diff
@@ -295,19 +295,20 @@ const TYPE = module.exports.TYPE = {
     writeParameterData: function(buffer, parameter, options) {
       if (parameter.value != null) {
         let days, dstDiff, milliseconds, seconds, threeHundredthsOfSecond;
+        var time = new Date(+parameter.value);
```
Comment (chdh) [inline src/data-type.js]: What is the rationale behind this conversion?
If `parameter.value` is already a `Date`, the `+` operator converts it to an integer and the `new Date()` converts it back to a `Date`. 
If `parameter.value` is a string that does not contain a number, the `+` operator produces a NaN.
e.g.: `+'2017-10-11'` ==> `NaN`
or: `+'Wed Oct 11 2017 19:49:46 GMT+0200'` ==> `NaN`

**Comment by chdh** [comment]
The root problem of this issue is that sometimes the parameter values are passed through `Type.validate()` (in `Request.validateParameters()`) and converted, but sometimes they are not passed through `Type.validate()` (in `BulkLoad.addRow()`).

A validation function like `Request.validateParameters()` should not have the side effect of converting the parameter values and `Type.validate()` should not return the converted value.

Parameter value conversion should be moved to a separate function which is then called from both `Type.validate()` and `Type.writeParameterData()`.

Another solution would be to rename `validate()` to e.g. `validateAndConvert()` and also call it in BulkLoad.

**Comment by atsage** [comment]
What file are you talking about?

**Comment by atsage** [comment]
So is this something that will be changed?

**Comment by chdh** [comment]
I think the whole conversion logic in the data-type.js module has to be revised.

Conversion to integer is also faulty. `parseInt()` is used even if the value is already a number and that can produce strange effects, e.g.:

`console.log(parseInt(0.000000006));`   ==> Output is 6!

Maybe it would be better not to convert at all and just throw a `TypeError` in `writeParameterData()` when the data type does not match? Why is it important to automatically convert a `string` to a `Date`?

**Comment by atsage** [comment]
I don't think that it's important to convert, the values should be passed in the correct format to begin with.  However, I've seen this issue a couple places and that is the fix.  It was initially posted here https://github.com/patriksimek/node-mssql/issues/185 and line 1413 of data-type.js has the same code to convert DATETIME2 time = new Date(+parameter.value);  I'm not a professional programmer, so if there is something that needs to be addressed in general with data-types.js then I would be glad to see it put in place.

**Comment by atsage** [comment]
After some more testing I have found that the problem is the node-mssql node converts the datetime into a number.  See here starting on line 166 https://github.com/patriksimek/node-mssql/blob/master/lib/datatypes.js  I will raise an issue there as well as this seems incorrect.

**Comment by atsage** [comment]
https://github.com/patriksimek/node-mssql/issues/546 Issue raised here

**Comment by chdh** [comment]
I don't agree. Starting at line 166 in https://github.com/patriksimek/node-mssql/blob/master/lib/datatypes.js, `Date` values are not converted to `number` but to a string in the format `N'yyyy-mm-dd hh:mm:ss'`.

**Comment by atsage** [comment]
You are right, I wasn't looking at it very hard.  It must get converted somewhere else to a number because when it gets to the section of this file it is indeed a number type.  Either way, neither of those types are a date and in node-mssql there is no possible way to send a date type so it fails.

**Comment by chdh** [comment]
@arthurschreiber @v-suhame what do you think about automatic value conversion for SQL statement parameters and BulkLoad field values? Should Tedious try to convert the data type of values that do not match, or just throw a TypeError?

Examples:

- Convert a string to an integer.
- Round a float value to an integer.
- Convert a string to a `Date`.

The current implementation converts values for SQL statement parameters but not for BulkLoad field values. There are some errors and discrepancies in `data-type.js` that should be fixed. How can we proceed?

**Comment by arthurschreiber** [comment]
> Maybe it would be better not to convert at all and just throw a TypeError in writeParameterData() when the data type does not match? Why is it important to automatically convert a string to a Date?

Yeah, so I totally think we should not be doing any conversion between datatypes. 👍 

> I don't think that it's important to convert, the values should be passed in the correct format to begin with.

Yup, again, this is the way to go. 😅 

---

I think long-term, here's what we need to do (and what probably is the only sane approach to this):

`tedious` should provide a system to plug in custom data-type support modules. Each module would handle parsing of TDS data values into JavaScript objects and converting JavaScript objects back into TDS data values.

Out-of-the box, `tedious` should only support TDS data types that can be converted to equivalent JavaScript data types without a loss of precision or other information. E.g. we would support "basic" data types like `varchar`, `char`, `binary`, `varbinary`, `int`, etc.

If a user of `tedious` needs to work with other data types that can't be represented in plain JavaScript (like `bigint`), they'd have to provide a data-type module. These could be either write by them (to satisfy their needs) or written by us (to provide a base set).

Let's use the `bigint` example again: There's currently a bunch of arbitrary precision integer libraries available for JavaScript (https://www.npmjs.com/browse/keyword/biginteger). Whichever one we pick, someone will be unhappy because we didn't use the library they use in their application. By having a pluggable system, we can provide a base set of types for "popular" libraries, and if they need something special, they can go and build their own.

This would also solve the issue of providing conversion logic - if someone believes they really need to be able to convert String objects to DateTime values, they should be able to hook up their own data-type handler and do whatever they want. 😄 

@chdh What do you think?

**Comment by chdh** [comment]
@arthurschreiber A plug-in architecture would be a very good and flexible solution. But in the short term, we should fix the problems with the "basic" data types. This would be a small change.

**Comment by arthurschreiber** [comment]
@chdh You're right, we need a fix in the interim and can't leave this broken. As you've worked on the BulkLoad code before, what would be you preferred way to fix this?

**Comment by chdh** [comment]
@arthurschreiber I would just clean up the existing `validate()` and `writeParameterData()` functions for each data type in data-type.js and add `validate()` calls to `BulkLoad`.

If we want backward compatibility, so that the PR can already be included in a Tedious 2.x release, I would add `convert()` functions for each data type and call these in `Request.validateParameters()` just after the `validate()` calls. In this case, `validateParameters()` should be renamed to `validateAndConvertParameters()`. We could add an option to enable/disable parameter conversion. For Tedious 2.x, the option would be enabled by default. For 3.x it would be disabled.

**Comment by niwsa** [comment]
Any update on the fix? I am facing the same issue.

**Comment by arthurschreiber** [comment]
Hey there,

I've been spending a lot of time going through the internals of tedious and it's data type system. There's a lot of weird code in there (as pointed out by @chdh).

I'll be opening a PR shortly that will fix this issue here by throwing a more descriptive error message when an unexpected type is passed to the different `writeParameterData` functions for different data types. As that is not a breaking change and will clearly signal to our users why they're seeing an error (which is much better than just throwing some obscure javascript error).

We'll have to find a better method of handling types long term, but that's a separate discussion from the error and fix proposed here. 😞 

Thank you for opening this PR and getting the discussion started. ❤️  I'm really sorry this has taken a long time to start moving to a solution 😞.

I really appreciate the time you spent and hope the solution I'm suggesting above is satisfactory. If not, please let us know!

**Comment by atsage** [comment]
Any progress is good, I'll take it!

**Comment by Shaffuan** [comment]
Any Progress Guy???

**Comment by esterkjuicepress** [comment]
Was this fix ever merged? I still have issues with bulk inserting DateTime types
```
TypeError: parameter.value.getTime is not a function
at Object.writeParameterData (\node_modules\tedious\lib\data-types\datetime.js:30:44)
```

**Comment by stegel** [comment]
was this ever resolved?

**Comment by thbl** [comment]
+1

**Comment by MichaelSun90** [comment]
@arthurschreiber, I had a different PR #907 that holds a similar change, and it already updated to the latest master.  I am not sure whether that we should close this PR and get the other PR merged in?

**Comment by arthurschreiber** [comment]
I think with the changes in https://github.com/tediousjs/tedious/pull/1065 and `validateBulkLoadParameters` set to `true` this is no longer necessary? Please feel free to reopen if you think I'm mistaken. 🙇‍♂️

**Pull request closed without merging**